### PR TITLE
Add SOFT_ASSERT to gracefully recover from invariant violations

### LINKED ISSUE #82229: [Profiler] Allow profiler to gracefully fail without interrupting workflow.
### 🚀 The feature, motivation and pitch

The pytorch profiler works very hard to minimize work on the hot path to reduce overhead. However, one of the consequences of this choice is that there is a fairly significant post processing step to convert these raw events into a coherent program representation. To ensure correctness this post processing makes liberal use of asserts. However this is causing a non-trivial number of post processing calls to fail and as a result crashes the entire workload. Which is not ideal. How we want to handle this depends on the context. In debug mode it should be a hard failure. In opt mode we should emit a warning (to console in OSS and to a remote logger in FBCode), record it in the trace file, and gracefully drop the offending event or do whatever else is needed to return to a somewhat sane state. (E.g. drop the event)

As a concrete example, we currently have a guard to ensure that an event produces sane timestamps before marking it finished and incorporating it in the tree:
```
void mark_finished(std::shared_ptr<Result>& r) {
  TORCH_INTERNAL_ASSERT(!r->finished_, r->name());
  r->finished_ = true;
  TORCH_INTERNAL_ASSERT(r->endTimeNS() >= r->start_time_ns_, r->name());
}
```
However corner cases (such as is asynchronously stopped when an op is in progress) breaks this. We want to correct the logic, but it's not necessary to crash an entire run. A better structure would be something like:
```
if (SOFT_ASSERT(!r->finished_, r->name()) {
  r->finished_ = true;
  return SOFT_ASSERT(r->endTimeNS() >= r->start_time_ns_, r->name());
}
return false;
```
And then the caller should act appropriately.

This will consist of two parts:
1) Implement `SOFT_ASSERT`. (Working title)
2) Update profiler to gracefully recover from invariant violations.

### Alternatives

_No response_

### Additional context

_No response_

cc @ilia-cher @robieta @chaekit @gdankel @bitfort @ngimel @nbcsm @guotuofeng @guyang3532 @gaoteng-git

### PR DESCRIPTION
**Pull request by davidchencsl**
Summary: Implement SOFT_ASSERT that only fails in debug mode, but only trigger a warning log in release mode. This allows us to gracefully handle some of the invariant violation when processing traces that doesn't necessarily need to crash the entire program.

Test Plan: Added SOFT_ASSERT test in containers.cpp

Differential Revision: D38327334

### CODE REVIEW

File: torch/csrc/profiler/util.h [hunk 1]
```diff
@@ -35,6 +35,21 @@
 #endif
 #endif
 
+// Define SOFT_ASSERT
+#ifdef NDEBUG
+#define HANDLE_ERROR(cond, ...) TORCH_WARN(__VA_ARGS__);
```
Comment (robieta) [inline torch/csrc/profiler/util.h:40]: This should be runtime toggle-able, no?

File: torch/csrc/profiler/util.h [hunk 2]
```diff
@@ -35,6 +35,24 @@
 #endif
 #endif
 
+// Define SOFT_ASSERT
+#ifdef NDEBUG
```
Comment (robieta) [inline torch/csrc/profiler/util.h:39]: As written, `NDEBUG` takes priority over the environment variable. I think it should be the other way around. (And of course that means the environment variable must be able to both force enable and force disable.) My reasoning is that I want to add unit test for the profiler that test various kinds of pathological conditions / invalid inputs and assert that they are handled gracefully. However if the behavior depends on the build flag then it's impossible to write unit tests that will always work.

Comment (davidchencsl) [reply torch/csrc/profiler/util.h:39]: Maybe let the env var disable the HANDLE_ERROR in both cases?

Comment (davidchencsl) [reply torch/csrc/profiler/util.h:39]: Ah I see what you are saying.

**Review by albanD — CHANGES_REQUESTED** [review]
I think this should be discussed in the context of https://github.com/pytorch/rfcs/pull/43 before adding a new API.

In particular, you could be raising a warning of a custom subclass and then use python warning filter to make this warning class an error if that's what you want for particular runs!

File: torch/csrc/profiler/util.h [hunk 3]
```diff
@@ -48,9 +65,9 @@ using steady_clock_t = std::conditional<
 inline time_t getTimeSinceEpoch() {
   auto now = std::chrono::system_clock::now().time_since_epoch();
   return std::chrono::duration_cast<std::chrono::nanoseconds>(now).count();
-}
+  }
```
Comment (robieta) [inline torch/csrc/profiler/util.h:68]: nit: unrelated formatting?

File: torch/csrc/profiler/util.h [hunk 4]
```diff
@@ -35,9 +36,25 @@
 #endif
 #endif
 
+// Define SOFT_ASSERT
+#define SOFT_ASSERT(cond, ...)                                     \
+  [&](const char* func, const char* file, uint32_t line) -> bool { \
```
Comment (robieta) [inline torch/csrc/profiler/util.h:41]: We should use these, since we went through the trouble of plumbing them in.

**Review by robieta — COMMENTED** [review]
Only major comment is to include the source info. (Tip: `c10::str` will help you package it, since you can't use fmt in a header.)

**Comment by robieta** [comment]
> I think this should be discussed in the context of [pytorch/rfcs#43](https://github.com/pytorch/rfcs/pull/43) before adding a new API.
> 
> In particular, you could be raising a warning of a custom subclass and then use python warning filter to make this warning class an error if that's what you want for particular runs!

Oh, interesting. I was not aware of this but it seems like exactly what we were going for. (Basically a warn that fails unit tests.)

**Comment by robieta** [comment]
@albanD What are your thoughts on landing this with the understanding that it's just a placeholder until we can move to https://github.com/pytorch/rfcs/pull/43. Because as awesome as that sounds, we do have a concrete need today. (And we can hide the APIs even more if you're concerned that they're not private enough.)

**Comment by albanD** [comment]
We plan on moving quite quickly on this, especially the part related to error and warning.

I think the only thing that is missing for you to be able to use the normal warning filter is for you to be able to warn with a custom subclass on the c++ side.
cc @kurtamohler that might be a good first step (c++ warning subclass and c++ warning filter) for the rfc implementation since we already have a need here haha

If you want to merge this before the warning is done, could you:
- Add a comment that this API will be changed soon once the rfc is done.
- Do you actually want to keep different behavior for debug vs non-debug build? I think this is the only thing that would prevent you from moving the API introduced here to use a warning-based system once it is available.

Impl sounds good to me otherwise.

**Comment by robieta** [comment]
> Add a comment that this API will be changed soon once the rfc is done.

Absolutely. This looks absolutely awesome and I'd be happy to work with @kurtamohler as a concrete early use case.

> Do you actually want to keep different behavior for debug vs non-debug build?

I would argue that an analog of `TORCH_INTERNAL_ASSERT_DEBUG_ONLY` would be useful. However it's not essential, and if it would slow things down I think we should drop it and just do what the rfc implementation can support. Better to move quickly to the idiomatic way ASAP and build from there.

File: torch/csrc/profiler/util.h [hunk 5]
```diff
@@ -35,9 +36,25 @@
 #endif
 #endif
 
+// Define SOFT_ASSERT
```
Comment (robieta) [inline torch/csrc/profiler/util.h:39]: // TODO: replace with https://github.com/pytorch/rfcs/pull/43 when it is ready.

**Review by robieta — APPROVED** [review]
LGTM. (Modulo Alban's request for a comment.)

**Comment by albanD** [comment]
@robieta quick question: Do you need the filtering system that can transform warnings into error to work in a c++-only world? Or is that ok if that is only available from python?

**Comment by robieta** [comment]
> @robieta quick question: Do you need the filtering system that can transform warnings into error to work in a c++-only world? Or is that ok if that is only available from python?

It's not ideal (we want to write more scoped C++ unit tests for C++ components), but not the end of the world.
